dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1070 @@
+     * or update the actual vartypes of the variables.
+     */
+    void change_vartype(Vartype source, Vartype target, index_type v) {
+        if (source == target) {
+            return;
+        } else if (source == Vartype::SPIN && target == Vartype::BINARY) {
+            if (has_adj()) {
+                for (auto it = (*adj_ptr_)[v].begin(); it != (*adj_ptr_)[v].end(); ++it) {
+                    linear_biases_[it->v] -= it->bias;
+                    (*adj_ptr_)[it->v][v] *= 2;  // log(n)
+                    it->bias *= 2;
+                }
+            }
+
+            offset_ -= linear_biases_[v];
+            linear_biases_[v] *= 2;
+        } else if (source == Vartype::BINARY && target == Vartype::SPIN) {
+            if (has_adj()) {
+                for (auto it = (*adj_ptr_)[v].begin(); it != (*adj_ptr_)[v].end(); ++it) {
+                    linear_biases_[it->v] += it->bias / 2;
+                    (*adj_ptr_)[it->v][v] /= 2;  // log(n)
+                    it->bias /= 2;
+                }
+            }
+
+            offset_ += linear_biases_[v] / 2;
+            linear_biases_[v] /= 2;
+        } else {
+            // todo: there are more we could support
+            throw std::logic_error("unsupported vartype change");
+        }
+    }
+
+    /**
+     * Adjust all of the biases in the model to reinterpret them as being
+     * from the source vartype to the target.
+     * 
+     * This method is meant to be used by subclasses, since it does not check
+     * or update the actual vartypes of the variables.
+     */
+    void change_vartypes(Vartype source, Vartype target) {
+        if (source == target) return;  // nothing to do
+
+        bias_type lin_mp, lin_offset_mp, quad_mp, quad_offset_mp, lin_quad_mp;
+        if (source == Vartype::SPIN && target == Vartype::BINARY) {
+            lin_mp = 2;
+            lin_offset_mp = -1;
+            quad_mp = 4;
+            lin_quad_mp = -2;
+            quad_offset_mp = .5;
+        } else if (source == Vartype::BINARY && target == Vartype::SPIN) {
+            lin_mp = .5;
+            lin_offset_mp = .5;
+            quad_mp = .25;
+            lin_quad_mp = .25;
+            quad_offset_mp = .125;
+        } else {
+            throw std::logic_error("unsupported vartype combo");
+        }
+
+        for (size_type u = 0; u < num_variables(); ++u) {
+            bias_type lbias = linear_biases_[u];
+
+            linear_biases_[u] *= lin_mp;
+            offset_ += lin_offset_mp * lbias;
+
+            if (has_adj()) {
+                for (auto it = (*adj_ptr_)[u].begin(); it != (*adj_ptr_)[u].end(); ++it) {
+                    bias_type qbias = it->bias;
+
+                    it->bias *= quad_mp;
+                    linear_biases_[u] += lin_quad_mp * qbias;
+                    offset_ += quad_offset_mp * qbias;
+                }
+            }
+        }
+    }
+
+    /// Resize model to contain n variables.
+    void resize(index_type n) {
```

Comment:
Is it to reduce memory space ?  Then after calling resize we should also call shrink_to_fit,  or actually resize using the swap function.

Reply:
No, not to reduce the space.